--- linux-2.6.12-rc2-orig/mm/swap.c	2005-04-04 18:38:07.000000000 +0200
+++ linux-2.6.12-rc2-fkt/mm/swap.c	2005-05-21 01:11:17.203976492 +0200
@@ -48,8 +48,11 @@ void put_page(struct page *page)
 		}
 		return;
 	}
-	if (!PageReserved(page) && put_page_testzero(page))
-		__page_cache_release(page);
+	if (test_bit(PG_FKT,&page->flags))
+		put_page_testzero(page);
+	else
+		if (!PageReserved(page) && put_page_testzero(page))
+			__page_cache_release(page);
 }
 EXPORT_SYMBOL(put_page);
 #endif
