--- linux-2.6.12-rc2-orig/include/linux/buffer_head.h	2005-04-04 18:40:04.000000000 +0200
+++ linux-2.6.12-rc2-fkt/include/linux/buffer_head.h	2005-05-21 01:12:00.946273325 +0200
@@ -145,6 +145,9 @@ void end_buffer_read_sync(struct buffer_
 void end_buffer_write_sync(struct buffer_head *bh, int uptodate);
 void end_buffer_async_write(struct buffer_head *bh, int uptodate);
 
+int buffers_dirty(struct page *);
+void wait_buffers_clean(struct page *);
+
 /* Things to do with buffers at mapping->private_list */
 void mark_buffer_dirty_inode(struct buffer_head *bh, struct inode *inode);
 int inode_has_buffers(struct inode *);
