--- linux-2.6.12-rc2-orig/include/linux/mm.h	2005-04-04 18:37:52.000000000 +0200
+++ linux-2.6.12-rc2-fkt/include/linux/mm.h	2005-05-21 01:10:55.129359256 +0200
@@ -341,8 +341,11 @@ static inline void get_page(struct page 
 
 static inline void put_page(struct page *page)
 {
-	if (!PageReserved(page) && put_page_testzero(page))
-		__page_cache_release(page);
+	if (test_bit(PG_FKT,&page->flags))
+		put_page_testzero(page);
+	else
+		if (!PageReserved(page) && put_page_testzero(page))
+			__page_cache_release(page);
 }
 
 #endif		/* CONFIG_HUGETLB_PAGE */
