--- linux-2.6.12-rc2-orig/kernel/sched.c	2005-04-04 18:39:17.000000000 +0200
+++ linux-2.6.12-rc2-fkt/kernel/sched.c	2005-05-21 01:11:46.433497295 +0200
@@ -18,6 +18,8 @@
  *  2004-04-02	Scheduler domains code by Nick Piggin
  */
 
+/***** 11-Oct-00	rdr	add fkt_always probe at switch_to *****/
+
 #include <linux/mm.h>
 #include <linux/module.h>
 #include <linux/nmi.h>
@@ -49,6 +51,9 @@
 #include <linux/acct.h>
 #include <asm/tlb.h>
 
+/* rdr */
+#include <linux/fkt.h>
+
 #include <asm/unistd.h>
 
 /*
@@ -1373,6 +1378,10 @@ task_t * context_switch(runqueue_t *rq, 
 	}
 
 	/* Here we just switch the register state and the stack. */
+
+	FKT_ALWAYS_PROBE1(FKT_KEYMASKALL, FKT_SWITCH_TO_CODE,          /* rdr */
+		next->pid);                                            /* rdr */
+
 	switch_to(prev, next, prev);
 
 	return prev;
