--- linux-2.6.12-rc2-orig/kernel/signal.c	2005-04-04 18:38:07.000000000 +0200
+++ linux-2.6.12-rc2-fkt/kernel/signal.c	2005-05-21 01:11:17.291963009 +0200
@@ -28,6 +28,8 @@
 #include <asm/unistd.h>
 #include <asm/siginfo.h>
 
+#include <linux/fkt.h>	/* st */
+
 /*
  * SLAB caches for signal bits.
  */
@@ -2249,6 +2251,7 @@ asmlinkage long
 sys_kill(int pid, int sig)
 {
 	struct siginfo info;
+	FKT_PROBE2(FKT_SYSTEM_KEYMASK, FKT_KILL_CODE, pid, sig); /* st */
 
 	info.si_signo = sig;
 	info.si_errno = 0;
